otlp/traces_direct.go

```diff
@@ -0,0 +1,1709 @@
+			}
+
+			// Note, the logic here will probably need to change once the Entity
+			// system is finalized.
+			err = unmarshalScopeSpans(ctx, slice, resourceAttrs, batch)
+			if err != nil {
+				return err
+			}
+
+		default:
+			// Note field 3 is schema_url, but we don't use it.
+			if err := skipField(data, &iNdEx, preIndex, l); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// unmarshalResource parses a Resource message
+func unmarshalResource(ctx context.Context, data []byte, attrs *msgpAttributes) error {
+	l := len(data)
+	iNdEx := 0
+
+	for iNdEx < l {
+		preIndex := iNdEx
+		fieldNum, wireType, err := decodeField(data, &iNdEx)
+		if err != nil {
+			return err
+		}
+
+		switch fieldNum {
+		case 1: // attributes
+			slice, err := decodeWireType2(data, &iNdEx, l, wireType)
+			if err != nil {
+				return err
+			}
+
+			// Parse KeyValue
+			err = unmarshalKeyValue(ctx, slice, attrs, 0)
+			if err != nil {
+				return err
+			}
+
+		default:
+			if err := skipField(data, &iNdEx, preIndex, l); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// parseKeyValue parses a KeyValue message and returns the key and raw value bytes
+func parseKeyValue(data []byte) ([]byte, []byte, error) {
+	var key []byte
+	var valueBytes []byte
+
+	l := len(data)
+	iNdEx := 0
+
+	for iNdEx < l {
+		preIndex := iNdEx
+		fieldNum, wireType, err := decodeField(data, &iNdEx)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		switch fieldNum {
+		case 1: // key
+			key, err = decodeWireType2(data, &iNdEx, l, wireType)
+			if err != nil {
+				return nil, nil, err
+			}
+
+		case 2: // value
+			valueBytes, err = decodeWireType2(data, &iNdEx, l, wireType)
+			if err != nil {
+				return nil, nil, err
+			}
+
+		default:
+			if err := skipField(data, &iNdEx, preIndex, l); err != nil {
+				return nil, nil, err
+			}
+		}
+	}
+
+	return key, valueBytes, nil
+}
+
+// processValueDirect handles a value recursively, tracking depth for proper flattening
+func processValueDirect(ctx context.Context, key []byte, value any, attrs *msgpAttributes, depth int) error {
+	var err error
+
+	// Handle different value types to match legacy behavior
+	switch v := value.(type) {
+	case []byte:
+		// Bytes are JSON encoded - match the legacy behavior
+		husky.AddTelemetryAttribute(ctx, "received_bytes_attr_type", true)
+		err = attrs.addAny(key, addAttributeToMapAsJsonDirect(v))
+		if err != nil {
+			return err
+		}
+	case []any:
+		// Arrays are JSON encoded
+		husky.AddTelemetryAttribute(ctx, "received_array_attr_type", true)
+		err = attrs.addAny(key, addAttributeToMapAsJsonDirect(v))
+		if err != nil {
+			return err
+		}
+	case map[string]any:
+		// Kvlists are flattened with dot notation
+		husky.AddTelemetryAttributes(ctx, map[string]any{
+			"received_kvlist_attr_type": true,
+			"kvlist_max_depth":          depth,
+		})
+		if depth < maxDepth {
+			// Flatten the kvlist
+			for k, v := range v {
+				flatKey := append(key, '.')
```

Comment:
This right here is the worst bug in this PR, and is why this kind of code is so dangerous. `key` is ultimately a reference to a subslice of the serialized input, so when you append to it you corrupt the input. And then you _probably_ get away with it, because deserialization has probably moved past the point you corrupt, and our generally doesn't read this data twice. But you might not always be so lucky.